File: utils/ocr_parsing_utils.js
const fs = require('fs')
const path = require('path')
const csv = require('csv-parser')
const H = require('highland')
const R = require('ramda')
const lunr = require('lunr')

var occupations = JSON.parse(fs.readFileSync("/Users/stephen/git/city-directories/utils/consolidated_occupations.json"));

/*
TODO - widow parsing doesn't seem to be working (5/5/2017)
TODO - improve handling of middle initials / company abbreviations (5/5/2017)
TODO - address cardinal direction scoring by looking at order of token (5/5/2017)
*/

/*
var ndjson_lines = fs.readFileSync("/Users/stephen/Documents/nypl-directories/1854-55/lines.ndjson", {
  'encoding': 'utf-8'
}).split("\n")
var interpreted_lines = []
ndjson_lines.forEach((line, index) => {
  console.log(index);
  parsed = JSON.parse(line);
  interpreted_lines.push([parsed, create_labeled_record(parsed['text'])]);
})
*/

/*
fs.writeFile("/Users/stephen/git/city-directories/utils/parsing_output/node_test_5k_5-5-2017.json", JSON.stringify(interpreted_lines,null,2))
*/

var job_idx = lunr(function() {
  this.ref('id')
  this.field('value')
  this.field('source')
  occupations.forEach(function(doc, index) {
    mdoc = doc;
    mdoc['id'] = index;
    this.add(doc)
  }, this)
})

function job_probability_score(token, search_index) {
  if (search_index.search(token).length > 0) {
    return 1.5;
  } else if (search_index.search(token + "~1").length > 0) {
    return 1.0;
  } else if (search_index.search(token + "~2").length > 0) {
    return 0.9;
  } else if (search_index.search(token + "~4").length > 0) {
    return 0.5;
  } else {
    return -0.5;
  }
}



var contents = fs.readFileSync("/Users/stephen/git/city-directories/utils/parsing_output/nyc-hpc_100k_5-1-2017/out_v85.json");
var json_content = JSON.parse(contents)

/* Class contains heuristic helper functions
   for labeling tokens */
class TokenInterpret {
  static contains_numbers(token) {
    return /\d/.test(token)
  }
  static matches_cardinal_dir(token) {
    return /^(s|S|w|W|e|E|n|N).{0,1}$/.test(token)
  }
  static no_white_space(token) {
    return !/^\S*(\s)+\S*$/.test(token)
  }
  static is_short(token) {
    return (token.length < 4)
  }
  static could_be_abbreviation(token, index) {
    return (token.length == 1 && index > 0)
  }
  static probably_job(token) {
    return true
  }
}

class InterpretedLine {
  constructor(original_ocr_object) {
    this.o_object = original_ocr_object;
    this.o_text = this.o_object['text']
  }
}

function tokenize(line) {
  t1 = line.split(/(,|\.\s)/).map((elem) => {
    return elem.trim();
  })
  t2 = t1.filter((elem) => {
    return !(/^\s*[,\.]\s*$/.test(elem))
  })
  t3 = []
  t2.forEach(function(token) {
    t3 = t3.concat(jitter_split(token));
  })
  return t3;
}

/* Split token into two if begins or ends with a single character */
function jitter_split(token) {
  if (token.slice(1, 2) == ' ' && token.slice(-2, -1) == ' ') {
    return [token.slice(0, 1), token.slice(2, -2), token.slice(-1)]
  } else if (token.slice(1, 2) == ' ') {
    return [token.slice(0, 1), token.slice(2)]
  } else if (token.slice(-2, -1) == ' ') {
    return [token.slice(0, -2), token.slice(-1)]
  } else {
    return [token]
  }
}

function category_vote(ordered_token_array) {
  decisions = ordered_token_array.map(((token, i) => {
    return {
      'token': token,
      'index': i,
      'votes': []
    }
  }));
  ordered_token_array.forEach(function(token, i) {
    if (TokenInterpret.is_short(token)) {
      if (TokenInterpret.contains_numbers(token)) {
        decisions[i]['votes'].push({
          'address_component': 1.0
        })
      } else if (TokenInterpret.matches_cardinal_dir(token)) {
        decisions[i]['votes'].push({
          'address_component': 1.0
        })
      } else {
        decisions[i]['votes'].push({
          'predicate': 1.0
        })
      }
    } else {
      // Not short
      if (TokenInterpret.contains_numbers(token)) {
        decisions[i]['votes'].push({
          'address_component': 2.0
        })
      }
      if (i == 0) {
        decisions[i]['votes'].push({
          'name_component': 1.0
        })
        if (!TokenInterpret.contains_numbers(token)) {
          decisions[i]['votes'].push({
            'name_component': 1.0
          })
        }
      } // End first token
      decisions[i]['votes'].push({
        'job_component': job_probability_score(token, job_idx)
      })
      if (TokenInterpret.no_white_space(token)) {
        decisions[i]['votes'].push({
          'address_component': 0.5
        })
      }
    }
  })
  return decisions;
}

function most_likely_class(token_decision_object_array) {
  return token_decision_object_array.map((entry) => {
    modified_entry = entry;
    sums = {
      'job_component': 0.0,
      'name_component': 0.0,
      'address_component': 0.0,
      'predicate': 0.0,
      'ambiguous': 0.0
    }
    entry['votes'].forEach(function(vote) {
      entries = Object.entries(vote)
      k = entries[0][0]
      v = entries[0][1]
      sums[k] += v
    })
    modified_entry['sums'] = sums
    return modified_entry;
  })
}

function winner_take_all(array_most_likely_classes) {
  return array_most_likely_classes.map((entry) => {
    modified_entry = entry;
    modified_entry['winning_class'] = Object.entries(entry['sums']).reduce((acc, val) => {
      return (val[1] > acc[1]) ? val : acc
    })
    return modified_entry;
  })
}

function recount_votes(token_decision_object) {
  modified_entry = token_decision_object;
  sums = {
    'job_component': 0.0,
    'name_component': 0.0,
    'address_component': 0.0,
    'predicate': 0.0,
    'ambiguous': 0.0
  }
  token_decision_object['votes'].forEach(function(vote) {
    entries = Object.entries(vote)
    k = entries[0][0]
    v = entries[0][1]
    sums[k] += v
  })
  modified_entry['sums'] = sums
  modified_entry['winning_class'] = Object.entries(modified_entry['sums']).reduce((acc, val) => {
    return (val[1] > acc[1]) ? val : acc
  })
  return modified_entry;
}

function semantic_tokenize(line) {
  return winner_take_all(most_likely_class(category_vote(tokenize(line))))
}

function create_labeled_record(line) {
  return consolidate_features(semantic_tokenize(line))
}

function previous_winning_class(decision_list, current_index) {
  if (current_index - 1 >= 0) {
    return decision_list[current_index - 1]['winning_class'][0]
  }
  return null;
}

function modify_probability_of_subsequent(list_of_acceptable, vote, decisions, curr_index) {
  if (curr_index + 1 < decisions.length) {
    next_decision = decisions[curr_index + 1]
    if (next_decision['winning_class'] != Object.keys(vote)[0] && list_of_acceptable.includes(next_decision['winning_class'][0])) {
      decisions[curr_index + 1].votes.push(vote);
      decisions[curr_index + 1] = recount_votes(decisions[curr_index + 1]);
      return [decisions[curr_index + 1], curr_index]
    } else {
      return modify_probability_of_subsequent(list_of_acceptable, vote, decisions, curr_index + 1);
    }
  }
  return false;
}

function subsequent_is_not_confident(decision_list, current_index) {
  if (current_index + 1 < decision_list.length) {
    next_decision = decision_list[current_index + 1]
    if (next_decision['winning_class'][1] < 1.0) {
      return next_decision['winning_class'][0];
    }
  }
  return null;
}

/* Returns a record with labeled attributes */
function consolidate_features(all_decisions) {
  record = {
    'subject': [],
    'location': []
  }
  all_decisions.forEach((token, index) => {
    parsed_class = token['winning_class'][0]
    token_value = token['token']
    if (index == 0 && parsed_class == 'name_component') {
      record['subject'].push({
        'value': token_value,
        'type': 'primary'
      })
    } else {
      switch (parsed_class) {
        case 'job_component':
          if (!record['subject'].length == 0) {
            record['subject'][0]['occupation'] = token_value;
          }
          break;
        case 'predicate':
          switch (token_value) {
            case 'wid':
              if (!record['subject'].length == 0) {
                record['subject'][0]['occupation'] = 'widow';
              }
              deceased_name = look_for_name_of_deceased(all_decisions, index)
              if (deceased_name) {
                record['subject'].push({
                  'value': deceased_name,
                  'type': 'deceased spouse of primary'
                })
              }
              break;
            case 'h':
              modify_probability_of_subsequent(['job_component', 'name_component'], {
                "address_component": 1.0
              }, all_decisions, index)
              attach_to_next(all_decisions, index, 'address_component', [{
                'type': 'home'
              }])
              break;
            case 'r':
              modify_probability_of_subsequent(['job_component', 'name_component'], {
                "address_component": 1.0
              }, all_decisions, index)
              attach_to_next(all_decisions, index, 'address_component', [{
                'position': 'rear'
              }])
              break;
          }
          break;
        case 'address_component':
          /* We check the confidence of the next token too,
            and may merge it into the address as well */
          subsequent_class = subsequent_is_not_confident(all_decisions, index)
          if (subsequent_class == "job_component") {
            all_decisions[index + 1]['winning_class'] = ['address_component', 1.0]
          }
          loc = {
            'value': token_value
          }
          if (token['additional']) {
            token['additional'].forEach((obj) => {
              pair = Object.entries(obj)[0]
              k = pair[0]
              v = pair[1]
              loc[k] = v
            })
          }
          mr = merge_if_directly_subsequent_is_alike(all_decisions, index, parsed_class)
          if (mr) {
            all_decisions[index + 1] = mr
          } else {
            record['location'].push(loc)
          }
          break;
      }
    }
  })
  return record;
}

function merge_if_directly_subsequent_is_alike(decision_list, current_index, current_token_class) {
  if (current_index + 1 < decision_list.length) {
    next_decision = decision_list[current_index + 1]
    if (next_decision['winning_class'][0] == current_token_class) {
      next_decision['token'] = decision_list[current_index]['token'] + " " + next_decision['token']
      if (decision_list[current_index]['additional']) {
        if (next_decision['additional']) {
          next_decision['additional'].concat(decision_list[current_index]['additional'])
        } else {
          next_decision['additional'] = decision_list[current_index]['additional']
        }
      }
      return next_decision;
    }
  }
  return false;
}

function attach_to_next(class_list, current_index, match_class, attributes) {
  if (current_index + 1 < class_list.length) {
    next_class = class_list[current_index + 1]
    if (next_class['winning_class'][0] == match_class) {
      attributes.forEach((att) => {
        if (next_class['additional']) {
          next_class['additional'].push(att)
        } else {
          next_class['additional'] = [att]
        }
      })
    } else {
      attach_to_next(class_list, current_index + 1, match_class, attributes)
    }
  }
}

function look_for_name_of_deceased(list_of_classes, current_index) {
  if (current_index + 1 < list_of_classes.count) {
    next_class = list_of_classes[current_index + 1]
    if (next_class['winning_class'][0] == 'name_component' || ['winning_class'][1] <= 0.5) {
      // we check if the next class is either a name_component, or had a low confidence
      next_class['winning_class'][0] = 'already_considered'
      return next_class['token']
    }
  }
  return null;
}


/*
var output = fs.createWriteStream('/Users/stephen/Documents/nypl-directories/1854-55/stream_out.ndjson')
var dest = H(fs.createWriteStream('/Users/stephen/Documents/nypl-directories/1854-55/stream_out.ndjson'))
labeled_records = []
H(fs.createReadStream('/Users/stephen/Documents/nypl-directories/1854-55/lines.ndjson')).split().compact().map(JSON.parse).each((line) => {
  labeled_records.push([line, create_labeled_record(line['text'])])
})
*/

File: utils/ocr_parsing_utils.rb
## sgb334@nyu.edu

####### README #########
########################

## 1. Load all of the functions into memory
## 2. Populate a list of occupations (i.e. jobs = [])
## 3. Parse NDJSON file, line by line, into a list
## 4. Tokenize each OCR line-record with parse_line_elements(line)
## 5. Perform category_parsing(line, jobs); algorithm votes for most
##      likely class of object
## 6. Time to count up all of the votes –– call classify_decisions()
##      to determine the winning class for each token
## 7. Take your list of labeled tokens and assemble them into an
##      actual record, using consolidate_classes()

## TODO: Add function for merging subsequent address tokens (4/26/17)
## TODO: Actually organize these functions into something coherent... (4/26/17)

#### Useful Procedures ####

require 'JSON'
require 'CSV'
lines = []
File.open('/Users/stephen/Downloads/1854-55/lines.ndjson').each do |line|
  lines << JSON.parse(line)
end

jobs = File.read('/Users/stephen/Documents/nypl-directories/occupations_data/ipums-occ-list.txt').split("\n")
jobs.concat(File.read('/Users/stephen/Documents/nypl-directories/occupations_data/wilson-occ-list.txt').split("\n"))
jobs.concat(["publisher", "realestate", "realestate", "greengrocer", "green grocer", "coachdriver"])

parsed_lines = lines.map { |line| [line, parse_line_elements(line)] }

lines_and_decisions = []

parsed_lines[20000..20200].each_with_index do |line, index|
  lines_and_decisions << [line[0], category_parsing(line[1], jobs)]
  if index % 100 == 0
    puts "#{index / parsed_lines.count}"
  end
end

levenshteyn_cache = {}

lines_and_classes = lines_and_decisions.map { |ent| [ent[0], classify_decisions(ent[1])] }

tester = lines_and_classes.map { |ent| consolidate_classes(ent[0], ent[1]) }

File.open("/Users/stephen/Documents/nypl-directories/sample_decisions_classes.json", "wb") do |f|
  f.write(JSON.pretty_generate(tester))
end

results = []
lines[20000..30000].each_with_index do |line,index|
  results << process_line(line, jobs, levenshteyn_cache)
  puts index
end

#### End Procedures ####

## Consolidates all of the below functions,
## and produces a labeled record from an OCR
## JSON line
def process_line(original_line, jobs, ld_cache)
  consolidate_classes(original_line, classify_decisions(category_parsing(parse_line_elements(original_line), jobs, ld_cache)))
end

## Turns list of classes into a merged record
def consolidate_classes(original_line, list_of_classes)
  record = {
      :original_ocr => original_line,
      :attributes_parsed => {
          :subject =>
              [
                  #{:value => "Curran Sarah", :type => "primary", :occupation => "widowed"},
                  #{:value => "Richard", :type => "widower of primary"}
              ],
          :location =>
              [
                  #{:value => "7 Sixth", :position => "rear", :type => "home"}
              ]
      }
  }

  list_of_classes.each_with_index do |classed_token, index|
    parsed_class = classed_token[1][0]
    value = classed_token[0]
    if index == 0 && parsed_class == :name_component
      record[:attributes_parsed][:subject] << {:value => value, :type => 'primary'}
    end
    if index > 0
      case parsed_class
        when :job_component
          unless record[:attributes_parsed][:subject].count < 1
            record[:attributes_parsed][:subject][0][:occupation] = value
          end
        when :predicate
          case value
            when "wid"
              unless record[:attributes_parsed][:subject].count < 1
                record[:attributes_parsed][:subject][0][:occupation] = 'widow'
              end
              deceased_name = look_for_name_of_deceased(list_of_classes,index)
              unless deceased_name.nil?
                record[:attributes_parsed][:subject] << {:value => deceased_name, :type => 'deceased spouse of primary'}
              end
              #attach_to_next(list_of_classes, index, :name_component, [{:type => 'deceased spouse of primary'}])
            when "h"
              attach_to_next(list_of_classes, index, :address_component, [{:type => 'home'}])
            when "r"
              attach_to_next(list_of_classes, index, :address_component, [{:position => 'rear'}])
            else
          end
        ## inner case
        when :address_component
          loc = {:value => value}
          classed_token[2..-1].each do |xtra_attr| ## add in any additional attributes from predicates
            xtra_attr.each do |k, v|
              loc[k] = v
            end
          end
          unless merge_if_directly_subsequent_is_alike(list_of_classes, index, classed_token)
            record[:attributes_parsed][:location] << loc
          end
        else
      end
    end ## indices after 0
  end ## loop of classes

  return record
end

def merge_if_directly_subsequent_is_alike(list_of_classes, current_index, current_classed_token)
  if current_index + 1 < list_of_classes.count
    next_class = list_of_classes[current_index + 1]
    if next_class[1][0] == current_classed_token[1][0] ## i.e. they are the same, address followed by address
      next_class[0] = "#{current_classed_token[0]} #{next_class[0]}"
      next_class.concat(current_classed_token[2..-1])
      return true
    end
  end
  return false
end

## If a 'wid' predicate is found, we make sure
## we can interpret the following token as a :name_component
## for the deceased
def look_for_name_of_deceased(list_of_classes, current_index)
  if current_index + 1 < list_of_classes.count
    next_class = list_of_classes[current_index + 1]
    if next_class[1][0] == :name_component || next_class[1][1] <= 0.5
       ## we check if the next class is either a name_component, or had a low confidence
       next_class[1][0] = :already_considered
       return next_class[0]
    end
  end
  return nil
end


## Usage: attach_to_next(:address_component, [{:position => 'rear'}])
def attach_to_next(class_list, current_index, match_class, attributes)
  if current_index + 1 < class_list.count
    next_class = class_list[current_index + 1]
    if next_class[1][0] == match_class
      attributes.each do |att|
        next_class << att
      end
    else
      attach_to_next(class_list, current_index + 1, match_class, attributes)
    end
  end
end

## Tokenize
def parse_line_elements(line)
  text_field = normalize_synonyms(line['text'])
  elements = text_field.split(/(,|\.\s)/).map { |x| x.strip }
  elements.map { |elem|
    unless /^\s*[,\.]\s*$/.match(elem).nil? then
      elements.delete(elem)
    end }
  tokens = []
  elements.map { |elem| tokens.concat(jitter_tokenizer(elem)) }
  return tokens
end

## Split token into two if begins or ends with a
## single character
def jitter_tokenizer(token)
  toReturn = []
  if (token[1] == " " && token[-2] == " ")
    toReturn << token[0..0]
    toReturn << token[2..-3]
    toReturn << token[-1..-1]
  elsif token[1] == " "
    toReturn << token[0..0]
    toReturn << token[2..-1]
  elsif token[-2] == " "
    toReturn << token[0..-3]
    toReturn << token[-1..-1]
  else
    toReturn << token
  end
  toReturn
end

## Manual string subset replacement
def normalize_synonyms(text_string)
  text_string.gsub(/\s(Av|Ave|av|ave)\.\s/, " Avenue ")
  #text_string.gsub(/\s(W)\.\s/, " West ")
  #ext_string.gsub(/\s(Av|Ave|av|ave)\.\s/, " Avenue ")
  #text_string.gsub(/\s(Av|Ave|av|ave)\.\s/, " Avenue ")
end

## Primary decision tree function
def category_parsing(ordered_array_of_tokens, jobs, ld_cache)
  decisions = ordered_array_of_tokens.each_with_index.map { |token, index| {:token => token, :index => index, :votes => []} }

  ordered_array_of_tokens.each_with_index do |token, index|
    if is_short?(token)
      #if could_be_abbreviation?(token, index)
      if contains_numbers?(token)
        decisions[index][:votes] << {:address_component => 1.0}
      elsif matches_cardinal_dir?(token)
        decisions[index][:votes] << {:address_component => 1.0}
      else
        decisions[index][:votes] << {:predicate => 1.0}
      end
      #else ## not an abbreviation
      #end
    else ## not short
      if contains_numbers?(token)
        decisions[index][:votes] << {:address_component => 1.0}
      end
      if index == 0 ## First token
        ## Automatically, this token is probably a name (based on position)
        decisions[index][:votes] << {:name_component => 1.0}
        unless contains_numbers?(token)
          ## If the first token doesn't contain numbers, even more likely that it is name
          decisions[index][:votes] << {:name_component => 1.0}
        end
      end ## end (first token)
      if ld_probably_job?(token, jobs, 2, ld_cache)
        decisions[index][:votes] << {:job_component => 1.0}
      else
        decisions[index][:votes] << {:job_component => -1.0}
      end
      if no_white_space?(token)
        decisions[index][:votes] << {:address_component => 0.5}
      end
    end # end (not short)
  end ## loop
  decisions
end

## Aggregate token decisions with the most likely token class
## We run this on a list of decisions for a single line entry
def classify_decisions(set_of_token_decisions)
  set_of_token_decisions.map { |token| [token[:token], most_likely_class(token)] }
end

## Used to combine two subsequent address components,
## or two subsequent job components, etc.,
## taking a loss-less join of attributes except for
## :value, which is concatenated, separated by single whitespace
def merge_components(c1, c2)

end

## Count up votes, winner takes all
def most_likely_class(decisions_for_single_token)
  sums = {
      :job_component => 0,
      :name_component => 0,
      :address_component => 0,
      :predicate => 0,
      :ambiguous => 0.1,
  }
  decisions_for_single_token[:votes].each do |vote|
    vote.each do |type, score|
      sums[type] += score
    end
  end
  sums.max_by { |k, v| v }
end

##################################
## Heuristic helper functions,
## used in decision tree
##################################

def contains_numbers?(token)
  !/\d/.match(token).nil?
end

def matches_cardinal_dir?(token)
  !/^(s|S|w|W|e|E|n|N).{0,1}$/.match(token).nil?
end

def no_white_space?(token)
  /^\S*(\s)+\S*$/.match(token).nil?
end

def is_short?(token)
  token.length < 4
end

def could_be_abbreviation?(token, index)
  token.length == 1 && index > 0
end

def guess_abbreviation(token)
end

## Note: good threshold is ~2?
def ld_probably_job?(token, jobs, threshold, levenshteyn_cache)
  match_report = closest_match(token, jobs, levenshteyn_cache)
  return match_report.keys[0] <= threshold
end

## Turn JSON lines into list of [original_line, parsed_tokens]
def create_parse_pairs(range_of_lines)
  range_of_lines.map { |line| [line, parse_line_elements(line)] }
end

## Given a list of tokens, return a map of token => count
def compute_frequencies(list)
  dict = {}
  list.each do |token|
    unless dict.has_key?(token)
      dict[token] = list.count(token)
    end
  end
  dict
end

## Complements the method above (turns frequences into a csv file)
def freqs_to_csv(frequencies, path)
  CSV.open(path, "wb") do |csv|
    frequencies.map { |k, v| csv << [k, v] }
  end
end

def parse_eval_csv(list_of_parse_pairs, path)
  CSV.open(path, "wb") do |csv|
    list_of_parse_pairs.map { |pair| csv << [pair[0]["text"], pair[1]] }
  end
end

## Find closest matching token from a library (using Levenshtein distance)
def closest_match(token, library, cache)
  if cache.has_key?(token)
    return cache[token]
  end
  closest_tokens = []
  closest_token_score = 999
  library.each do |entity|
    ld_score = levenshtein_distance(token, entity)
    if ld_score < closest_token_score
      closest_token_score = ld_score
      closest_tokens = [entity]
    elsif ld_score == closest_token_score
      closest_tokens << entity
    end
  end
  cache[token] = {closest_token_score => closest_tokens}
  {closest_token_score => closest_tokens}
end

##################################
## Miscellaneous utility functions
##################################

## Credit: https://stackoverflow.com/questions/16323571/measure-the-distance-between-two-strings-with-ruby
def levenshtein_distance(s, t)
  m = s.length
  n = t.length
  return m if n == 0
  return n if m == 0
  d = Array.new(m+1) { Array.new(n+1) }

  (0..m).each { |i| d[i][0] = i }
  (0..n).each { |j| d[0][j] = j }
  (1..n).each do |j|
    (1..m).each do |i|
      d[i][j] = if s[i-1] == t[j-1] # adjust index into string
                  d[i-1][j-1] # no operation required
                else
                  [d[i-1][j]+1, # deletion
                   d[i][j-1]+1, # insertion
                   d[i-1][j-1]+1, # substitution
                  ].min
                end
    end
  end
  d[m][n]
end
